src/ops/custom.rs

```diff
@@ -1,210 +1,157 @@
 //! Extensible operations.
 
-use downcast_rs::{impl_downcast, Downcast};
 use smol_str::SmolStr;
-use std::any::Any;
-use std::cell::OnceCell;
 use std::collections::HashMap;
-use std::ops::Deref;
+use std::rc::Rc;
 
-use crate::hugr::Hugr;
-use crate::macros::impl_box_clone;
-use crate::resource::ResourceSet;
-use crate::types::SimpleType;
-use crate::types::{Signature, SignatureDescription};
+use crate::hugr::{HugrMut, HugrView};
+use crate::resource::{OpDef, ResourceId, ResourceSet};
+use crate::types::{type_arg::TypeArgValue, Signature, SignatureDescription};
+use crate::{Hugr, Resource};
 
-/// A wrapped [`CustomOp`] with fast equality checks.
-#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
-pub struct OpaqueOp {
-    /// Operation name, cached for fast equality checks.
-    id: SmolStr,
+use super::tag::OpTag;
+use super::{LeafOp, OpName, OpTrait, OpType};
 
-    /// The custom operation.
-    custom: Box<dyn CustomOp>,
+/// An instantiation of an [`OpDef`] with values for the type arguments
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+#[serde(into = "OpaqueOp", from = "OpaqueOp")]
+pub enum ExternalOp {
+    Resource {
+        def: Rc<OpDef>,
```

Comment:
I don't think we want to lose `Send` on the `Hugr`. We probably should use an `Arc` here.

In the future we _could_ add a lifetime parameter to the Hugr and make this a reference... but that would take some work.

Reply:
@aborgna-q thanks, `Rc` vs `Arc` was something else I meant to bring up that got lost among the other discussions. Will do.
Is making that change sufficient to restore `Send`ness? (And, why isn't there a compile-time error if we've just lost that...can we add one?)

Reply:
We should probably make some change to enforce `Send`ability in a preliminary PR, given this one is still being discussed!

Reply:
Yes, `Arc` is `Send + Sync`.
There's nothing requiring `Send` on `Hugr` at the moment, perhaps we could have a simple test to check for it.

Reply:
So there's example code for a test here: https://github.com/rust-lang/rust/issues/44363
But rustdoc at least is able to work this out automagically: https://github.com/rust-lang/rust/issues/17606
The latter thus seems preferable, but how to turn this into a CI test? Scanning/parsing the generated docs sounds worse than writing the code snippet in the first of those issues.

Reply:
In our case we can add a "test" that fails to compile if the traits are missing:

(in `hugr.rs`)
```rust

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn impls_send_sync() {
        trait Test: Send + Sync {}
        impl Test for Hugr {}
    }
}
```